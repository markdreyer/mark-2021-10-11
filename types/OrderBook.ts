export type OrderFeed = [number, number]
export enum OrderSide {
    BID = 'bid',
    ASK = 'ask',
}
export interface Order {
    price: number
    size: number
    total: number
}
