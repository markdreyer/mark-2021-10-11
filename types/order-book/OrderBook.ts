// lodash seems to perform better than native with sorted data
// https://www.measurethat.net/Benchmarks/ShowResult/228038
import _findIndex from 'lodash/findIndex'
import _sortedIndexBy from 'lodash/sortedIndexBy'
import { flow } from 'lodash/fp'
import { round } from '../../formats'

export type OrderFeed = [number, number]
export type OrderSide = 'bid' | 'ask'
export interface Order {
    price: number
    size: number
    total: number
}

export class OrderBook {
    public levelsDeep = 10
    public bids: Order[] = []
    public asks: Order[] = []
    readonly spread = () => this.topAsk() - this.topBid()
    readonly spreadPercent = () => round((this.spread() / this.topBid()) * 100)

    private topBid = (): number => this.bids[0]?.price
    private topAsk = (): number => this.asks.slice(-1)[0]?.price

    //asks ascending
    //bids descending
    constructor(bids: OrderFeed[], asks: OrderFeed[]) {
        this.processFeed(bids, asks)
    }
    public processFeed = (bids: OrderFeed[], asks: OrderFeed[]) =>
        flow(
            this.applyOrders(bids, 'bid'),
            this.applyOrders(asks, 'ask'),
            () => this
        )(this)

    public applyOrders =
        (feed: OrderFeed[], side: OrderSide) => (book: OrderBook) =>
            flow(
                book.mapOrders,
                book.upsert(side),
                book.filterZeroSizes,
                book.sumSizes(side),
                book.trimOrders(side),
                (orders) => {
                    switch (side) {
                        case 'bid':
                            book.bids = [...orders]
                            break
                        case 'ask':
                            book.asks = [...orders]
                            break
                    }
                },
                () => book
            )(feed, side)

    private mapOrders = (feed: OrderFeed[], side: OrderSide): Order[] =>
        feed.map(this.mapOrder)

    private mapOrder = (feed: OrderFeed): Order => ({
        price: feed[0],
        size: feed[1],
        total: 0,
    })
    private upsert = (side: OrderSide) => (newOrders: Order[]) =>
        this.upsertSorted(side, newOrders)

    private sumSizes =
        (side: OrderSide) =>
        (orders: Order[]): Order[] => {
            switch (side) {
                case 'bid':
                    for (let i = 0; i < orders.length - 1; i++) {
                        let order = orders[i],
                            prev = orders[i - 1]
                        order.total =
                            i === 0 ? order.size : prev.total + order.size
                    }
                    return orders
                case 'ask':
                    for (let i = orders.length - 1; i >= 0; i--) {
                        let order = orders[i],
                            prev = orders[i + 1]
                        order.total =
                            i === orders.length - 1
                                ? order.size
                                : prev.total + order.size
                    }
                    return orders
                default:
                    return orders
            }
        }
    private filterZeroSizes = (orders: Order[]): Order[] =>
        orders.filter((o) => o.size > 0)

    private trimOrders =
        (side: OrderSide) =>
        (orders: Order[]): Order[] =>
            side === 'bid'
                ? orders.slice(0, this.levelsDeep)
                : orders.slice(-this.levelsDeep)

    /**
     * BIDS are sorted DESCENDING
     * ASKS are sorted ASCENDING
     * We use this to iterate
     * different directions based on the side
     */
    private upsertSorted = (side: OrderSide, newOrders: Order[]): Order[] =>
        newOrders.reduce(
            (acc: Order[], order: Order) => {
                let updateIndex = _findIndex(
                    acc,
                    (o: Order) => o?.price === order.price
                )

                if (updateIndex !== -1) {
                    acc[updateIndex].size = order.size
                } else {
                    let insertIndex = _sortedIndexBy(acc, order, (o: Order) =>
                        side === 'bid' ? -o.price : o.price
                    )
                    acc.splice(insertIndex, 0, order)
                }
                return acc
            },
            side === 'bid' ? this.bids : this.asks
        )
}
