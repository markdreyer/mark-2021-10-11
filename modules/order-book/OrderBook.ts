// lodash seems to perform better than native with sorted data
// https://www.measurethat.net/Benchmarks/ShowResult/228038
import _findIndex from 'lodash/findIndex'
import _sortedIndexBy from 'lodash/sortedIndexBy'
import { flow, head, last } from 'lodash/fp'
import { round } from '../../formats'
import { Order, OrderFeed, OrderSide } from '../../types/OrderBook'
import { config } from '../../config'

export class OrderBook {
    public levelsDeep: number = config.levelsDeep
    public bids: Order[] = []
    public asks: Order[] = []
    private topAsk = () => last(this.asks)?.price || 0
    private topBid = () => head(this.bids)?.price || 0
    private hasSpread = () => this.bids.length > 0 && this.asks.length > 0
    public maxQuantity = 0

    readonly spread = (): number => this.hasSpread() ? round(this.topAsk() - this.topBid()) : 0
    readonly midPoint = (): number => this.topAsk() + this.topBid() / 2
    readonly spreadPercent = (): number => round((this.spread() / this.midPoint()) * 100) || 0

    constructor(bids: OrderFeed[], asks: OrderFeed[]) {
        this.processFeed(bids, asks)
    }
    public processFeed = (bids: OrderFeed[], asks: OrderFeed[]): OrderBook => {
        this.maxQuantity = 0
        flow(
            this.processOrders(bids, OrderSide.BID),
            this.processOrders(asks, OrderSide.ASK),
            () => this
        )()

        this.bids = this.calculatePercentOfBook(this.bids)
        this.asks = this.calculatePercentOfBook(this.asks)

        return this
    }

    private processOrders = (feed: OrderFeed[], side: OrderSide) => () =>
        flow(
            this.mapOrders,
            this.upsert(side),
            this.filterEmptyOrders,
            this.sumOrders(side),
            this.trimBook(side),
            this.commitOrders(side),
            () => this
        )(feed)

    private mapOrders = (feed: OrderFeed[]): Order[] => feed.map(this.mapOrder)

    private mapOrder = (feed: OrderFeed): Order => ({
        price: feed[0],
        size: feed[1],
        total: 0,
        percentOfBook: ''
    })

    private sumOrders = (side: OrderSide) => (orders: Order[]): Order[] => {
        return side === OrderSide.BID
            ? orders.reduce((acc, order, i) => {
                acc[i].total =
                    i === 0 ? order.size : acc[i - 1].total + order.size
                this.maxQuantity = Math.max(this.maxQuantity, acc[i].total)
                return acc
            }, orders)
            : orders.reduceRight((acc, order, i) => {
                acc[i].total =
                    i === acc.length - 1
                        ? order.size
                        : acc[i + 1].total + order.size
                this.maxQuantity = Math.max(this.maxQuantity, acc[i].total)
                return acc
            }, orders)
    }

    private calculatePercentOfBook = (orders: Order[]) =>
        orders.reduce((acc, order, i) => {
            acc[i].percentOfBook = round(order.total / this.maxQuantity * 100) + '%'
            return acc
        }, orders)

    private filterEmptyOrders = (orders: Order[]): Order[] =>
        orders.filter((o) => o.size > 0)

    private trimBook = (side: OrderSide) => (orders: Order[]): Order[] =>
        side === OrderSide.BID
            ? orders.slice(0, this.levelsDeep)
            : orders.slice(-this.levelsDeep)

    private upsert = (side: OrderSide) => (newOrders: Order[]): Order[] =>
        newOrders.reduce(
            (acc: Order[], order: Order) => {
                const updateIndex = _findIndex(acc, (o: Order) => o?.price === order.price)

                if (updateIndex !== -1) {
                    // Update
                    acc[updateIndex].size = order.size
                } else {
                    // Insert (while maintaining sort order)
                    const insertIndex = _sortedIndexBy(acc, order, (o: Order) => -o.price)
                    acc.splice(insertIndex, 0, order)
                }
                return acc
            },
            side === OrderSide.BID ? this.bids : this.asks
        )

    private commitOrders = (side: OrderSide) => (orders: Order[]) =>
        side === OrderSide.BID ? (this.bids = orders) : (this.asks = orders)
}
