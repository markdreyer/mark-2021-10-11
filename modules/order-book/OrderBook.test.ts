import { OrderBook, OrderFeed } from './OrderBook'

const asks: OrderFeed[] = [
    [30000, 200],
    [21000, 200],
]
const bids: OrderFeed[] = [
    [20000, 150],
    [15000, 150],
]

test('snapshot', () => {
    const actual = new OrderBook(bids, asks)
    expect(actual).toMatchSnapshot()
})
test('totals are sorted descending towards spread', () => {
    const actual = new OrderBook(bids, asks)
    actual.processFeed([[1, 100]], [[50000, 100]])

    expect(actual.bids.map((b) => b.total)).toStrictEqual([150, 300, 400])
    expect(actual.asks.map((b) => b.total)).toStrictEqual([500, 400, 200])
})
test('prices are sorted properly', () => {
    const actual = new OrderBook(bids, asks)
    actual.processFeed([[10000, 100]], [[50000, 100]])

    expect(actual.bids.map((b) => b.price)).toStrictEqual([20000, 15000, 10000])
    expect(actual.asks.map((b) => b.price)).toStrictEqual([50000, 30000, 21000])
})
test('spread', () => {
    const actual = new OrderBook(bids, asks)
    expect(actual.spread()).toBe(1000)
})
test('spread with empty orders', () => {
    const emptyAsk = new OrderBook(bids, [])
    const emptyBid = new OrderBook([], asks)
    const emptyBook = new OrderBook([], [])

    expect(emptyAsk.spread()).toBe(0)
    expect(emptyBid.spread()).toBe(0)
    expect(emptyBook.spread()).toBe(0)
    expect(emptyBook.spreadPercent()).toBe(0)
})
test('spread percent', () => {
    const actual = new OrderBook(bids, asks)
    actual.processFeed([[20502.5, 1000]], [])
    expect(actual.spreadPercent()).toBe(1.59)
})
test('apply deltas unordered', () => {
    const asks: OrderFeed[] = [
        [47274, 200],
        [47363.5, 200],
    ]
    const bids: OrderFeed[] = [
        [47258.5, 150],
        [47224.5, 150],
    ]
    const askDeltas: OrderFeed[] = [
        [47274, 0],
        [47275, 10],
        [47362, 25],
        [47363.5, 700],
    ]

    const bidDeltas: OrderFeed[] = [
        [47225.5, 100],
        [47258.5, 0],
        [47268.5, 10],
        [47224.5, 100],
    ]
    const actual = new OrderBook(bids, asks)
    actual.processFeed(bidDeltas, askDeltas)

    expect(actual.bids.reduce((acc, curr) => acc + curr.size, 0)).toBe(210)
    expect(actual.asks.reduce((acc, curr) => acc + curr.size, 0)).toBe(735)
    expect(actual.bids.map((b) => b.price)).toStrictEqual([
        47268.5, 47225.5, 47224.5,
    ])
    expect(actual.asks.map((b) => b.price)).toStrictEqual([
        47363.5, 47362, 47275,
    ])
})

test('empty deltas', () => {
    let actual = new OrderBook(bids, asks)
    actual = actual.processFeed([], [])

    expect(actual.bids.length).toBe(bids.length)
    expect(actual.asks.length).toBe(asks.length)
})

test('levels deep 1', () => {
    const book = new OrderBook(bids, asks)
    book.levelsDeep = 1

    book.processFeed(
        [
            [40000, 100],
            [40000, 200],
        ],
        []
    )
    book.processFeed(
        [[40000, 200]],
        [
            [50000, 200],
            [50001, 0],
        ]
    )

    expect(book.bids.length).toBe(1)
    expect(book.asks.length).toBe(1)
})
test('levels deep 2', () => {
    const book = new OrderBook(bids, asks)
    book.levelsDeep = 2

    book.processFeed(
        [
            [40000, 100],
            [40000, 200],
        ],
        []
    )
    book.processFeed(
        [[40000, 200]],
        [
            [50000, 200],
            [50001, 0],
        ]
    )

    expect(book.bids.length).toBe(2)
    expect(book.asks.length).toBe(2)
})
test('highest quantity on ask', () => {
    const book = new OrderBook(bids, asks)
    expect(book.maxTotal()).toBe(400)
})
test('highest quantity on bid', () => {
    const book = new OrderBook(bids, asks)
    book.processFeed([[40000, 101]],[])
    expect(book.maxTotal()).toBe(401)
})
