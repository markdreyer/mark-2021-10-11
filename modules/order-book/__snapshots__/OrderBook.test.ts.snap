// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
OrderBook {
  "asks": Array [
    Object {
      "percentOfBook": "100%",
      "price": 30000,
      "size": 200,
      "total": 400,
    },
    Object {
      "percentOfBook": "50%",
      "price": 21000,
      "size": 200,
      "total": 200,
    },
  ],
  "bids": Array [
    Object {
      "percentOfBook": "37.5%",
      "price": 20000,
      "size": 150,
      "total": 150,
    },
    Object {
      "percentOfBook": "75%",
      "price": 15000,
      "size": 150,
      "total": 300,
    },
  ],
  "calculatePercentOfBook": [Function],
  "commitOrders": [Function],
  "filterEmptyOrders": [Function],
  "hasSpread": [Function],
  "levelsDeep": 12,
  "mapFeedToOrder": [Function],
  "maxQuantity": 400,
  "midPoint": [Function],
  "process": [Function],
  "processFeed": [Function],
  "spread": [Function],
  "spreadPercent": [Function],
  "sumOrders": [Function],
  "toOrder": [Function],
  "topAsk": [Function],
  "topBid": [Function],
  "trimBook": [Function],
  "upsert": [Function],
}
`;
